es-ds-components/components/es-segmented-control.vue

```diff
@@ -0,0 +1,237 @@
+}
+
+interface ButtonDimensionsInterface {
+    options: Partial<CSSStyleDeclaration>[];
+}
+
+const props = withDefaults(defineProps<Props>(), {
+    disabled: false,
+});
+const model = defineModel<number>();
+const segmentedControlRef = useTemplateRef('segmentedControl');
+let resizeObserver: ResizeObserver | null = null;
+
+// needed to avoid Content Layout Shift (CLS)
+// as the inkbar is positioned via left and width attributes which are not set until mount
+// meaning the active state would be initially missing and then pop in on mount, causing CLS
+const isMounted = ref(false);
+
+const buttonDimensions: ButtonDimensionsInterface = reactive({
+    options: props.options?.length ? props.options.map(() => ({})) : [],
+});
+
+const inkbarStyle: ComputedRef<Partial<CSSStyleDeclaration>> = computed(
+    () => buttonDimensions.options[model.value || 0],
+);
+
+// assume the provided label is unique and convert to an id
+// by lowercasing and replacing all whitespace with dashes
+const labelId = computed(() => {
+    return props.label ? props.label.toLowerCase().replace(/\s/g, '-') : '';
+});
+
+// return the left/width to use for the active state bubble, for each provided button
+const getButtonDimensions = (buttons: HTMLCollection) => {
+    const result = [];
+    let runningLeftTotal = 0;
+    for (let i = 0; i < buttons.length; i += 1) {
+        const button = buttons[i] as unknown as HTMLElement;
+        result.push({
+            left: `${runningLeftTotal}px`,
+            width: `${button.offsetWidth}px`,
+        });
+
+        runningLeftTotal += button.offsetWidth;
+    }
+
+    return result;
+};
+
+onMounted(() => {
+    isMounted.value = true;
+
+    if (segmentedControlRef.value) {
+        // set the initial button dimensions
+        const buttons = segmentedControlRef.value.querySelectorAll(
+            '.es-segmented-control-button',
+        ) as unknown as HTMLCollection;
+        buttonDimensions.options = getButtonDimensions(buttons);
+
+        // set up a resize observer for whenever the segmented control's button group changes width
+        // this will most commonly happen when changing between desktop and mobile breakpoints
+        // but could also happen if the label text of any options changes
+        resizeObserver = new ResizeObserver((groups) => {
+            groups.forEach((group) => {
+                const updatedButtons = group.target.children;
+                buttonDimensions.options = getButtonDimensions(updatedButtons);
+            });
+        });
+
+        // start the resize observer listening to the segmented control's button group
+        const group = segmentedControlRef.value.querySelector('.es-segmented-control-group');
+        if (group) {
+            resizeObserver.observe(group);
+        }
+    }
+});
+
+onUnmounted(() => {
+    if (resizeObserver) {
+        resizeObserver.disconnect();
+    }
+});
+</script>
+
+<template>
+    <div
+        v-if="$props.options?.length"
```

Comment:
Nit: we don't need to use `$props.` or `props.` in the templates when accessing props. It's sufficient to say `v-if="length"`.